logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,209 @@
+                    }
+                }
+                byte[] marker = retrieveSecret(logstashMarker);
+                if (marker == null) {
+                    throw new SecretStoreException.NotLogstashKeyStore(String.format("Found a keystore at %s, but it is not a Logstash keystore.",
+                            keyStorePath.toAbsolutePath().toString()));
+                }
+                LOGGER.debug("Using existing keystore at {}", keyStorePath.toAbsolutePath());
+            } catch (NoSuchFileException noSuchFileException) {
+                LOGGER.info("Keystore not found at {}. Creating new keystore.", keyStorePath.toAbsolutePath());
+                //wrapped in a system property mostly for testing, but could be used for more restrictive defaults
+                String keyStorePermissions = System.getProperty("logstash.keystore.file.perms", "rw-rw----");
+                //create the keystore on disk with a default entry to identify this as a logstash keystore
+                Files.createFile(keyStorePath, PosixFilePermissions.asFileAttribute(PosixFilePermissions.fromString(keyStorePermissions)));
+                try (final OutputStream os = Files.newOutputStream(keyStorePath, StandardOpenOption.WRITE)) {
+                    keyStore = KeyStore.Builder.newInstance(KEYSTORE_TYPE, null, protectionParameter).getKeyStore();
+                    SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+                    byte[] base64 = SecretStoreUtil.base64Encode(LOGSTASH_MARKER.getBytes(StandardCharsets.UTF_8));
+                    SecretKey secretKey = factory.generateSecret(new PBEKeySpec(SecretStoreUtil.asciiBytesToChar(base64)));
+                    keyStore.setEntry(logstashMarker.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+                    keyStore.store(os, this.keyStorePass);
+                    PosixFileAttributeView attrs = Files.getFileAttributeView(keyStorePath, PosixFileAttributeView.class);
+                    if (attrs != null) {
+                        //the directory umask applies when creating the file, so re-apply permissions here
+                        attrs.setPermissions(PosixFilePermissions.fromString(keyStorePermissions));
+                    }
+                    LOGGER.info("Keystore created at {}", keyStorePath.toAbsolutePath());
+                } catch (Exception e) {
+                    throw new SecretStoreException.CreateException("Failed to create Logstash keystore.", e);
+                }
+            }
+        } catch (SecretStoreException sse) {
+            throw sse;
+        } catch (Exception e) { //should never happen
+            throw new SecretStoreException("Error while trying to create or load the Logstash keystore", e);
+        }
+    }
+
+    @Override
+    public Collection<SecretIdentifier> list() {
+        Set<SecretIdentifier> identifiers = new HashSet<>();
+        try {
+            readLock.lock();
```

Comment:
retrieveSecret and purgeSecret both load the keystore (open an inputstream, call keystore.load, etc) at the beginning of the method, but this one does not. Expected?

Reply:
Nice catch! ... yes this indeed a bug. We need to reload the keystore from disk to ensure we have an updated view of the keystore. This is is also an issue for persist. Will fix with tests.